### PR TITLE
Disallow concurrent queries within a transaction

### PR DESCRIPTION
**Pull request by jessesleeping**
- Detect concurrent queries in SqlQueryManager and throw an exception.
- Refactor the test cases and add a test case for concurrent queries.
- When generate a new transaction for test usage in TransactionBuilder, we mark it as inactive before executing the next query.

#7564

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 1]
```diff
@@ -349,6 +350,13 @@ public QueryInfo createQuery(SessionSupplier sessionSupplier, String query)
                     throw new PrestoException(NOT_SUPPORTED, "EXPLAIN ANALYZE only supported for statements that are queries");
                 }
             }
+            if (session.getTransactionId().isPresent()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: We should move this to `TransactionManager.setActive()`:
```java
idleStartTime.getAndUpdate(value -> {
    if (value == null) {
        throw ... INVALID_CONCURRENT_QUERY;
    }
    return null;
});

**Comment by jessesleeping** [comment]
@electrum @raghavsethi I failed some tests because when we create a transaction in the test cases, we don't mark the transaction as inactive before we run the actual query. I am considering marking a transaction as inactive in TransactionMeta's constructor. This will make the logic of concurrent query detection clearer and make things easier when testing transaction related components. 

Any comments?

**Comment by jessesleeping** [comment]
I didn't change the TransactionMetadata's constructor in the updated PR. Worked around the problem by manually setting the transaction as inactive in the failed test cases.

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
+/*
```
Comment (raghavsethi) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: Remove the period at the end of the commit message.

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import static com.facebook.presto.execution.QueryState.RUNNING;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class TestQueryRunnerUtil
+{
+    private TestQueryRunnerUtil() {}
+
+    public static final String LONG_LASTING_QUERY = "SELECT COUNT(*) FROM lineitem";
+    public static final String START_TXN_QUERY = "START TRANSACTION";
```
Comment (raghavsethi) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: This is new functionality, and should be in the second commit.

**Comment by jessesleeping** [comment]
``com.facebook.presto.execution.resourceGroups.db.TestQueues.TestBasic`` is not deterministic. This [assertion](https://github.com/prestodb/presto/blob/3e7fb92bbbb65dda7ae60d3c4edc635c077e87ad/presto-tests/src/test/java/com/facebook/presto/execution/resourceGroups/db/TestQueues.java#L99) assumes that the long lasting transaction is still running. It's possible that on this assertion the first query has already finished, which causes the test to fail.

**Comment by raghavsethi** [comment]
So in the current version of your PR, is there a flaky test?

**Comment by jessesleeping** [comment]
Yes.. Actually all test cases that use the ``LONG_LASTING_QUERY`` can be flaky tests... The one I mentioned above is more sensitive since it assumes the query last for more than 2 seconds. I think this is a separate issue. We may need to implement a mocked ``LONG_LASTING_QUERY`` which stays active unless canceled.

**Comment by jessesleeping** [comment]
I have created an issue for the flaky tests (#7667 ). To avoid conflicts, I will send another PR to fix those tests after this one is merged.

**Comment by jessesleeping** [comment]
After some investigation, I think #7667 is unrelated to this PR. Only the db/TestQueues.java needs to be changed. I will submit a separate PR for #7667 soon.

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueues.java [hunk 4]
```diff
@@ -64,40 +57,40 @@ private void testBasic(boolean resourceGroups)
         }
         Map<String, String> properties = builder.build();
 
-        try (DistributedQueryRunner queryRunner = createQueryRunner(properties)) {
+        try (DistributedQueryRunner queryRunner = TestQueryRunnerUtil.createQueryRunner(properties)) {
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueues.java]: Static import all of these helper methods

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java (hunk 2 above)
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: Typo "utlility" in commit message

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class TestQueryRunnerUtil
+{
+    private TestQueryRunnerUtil() {}
+
+    public static final String LONG_LASTING_QUERY = "SELECT COUNT(*) FROM lineitem";
+
+    public static QueryId createQuery(DistributedQueryRunner queryRunner, Session session, String sql)
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: Are there any changes to this code, or is it purely a move of the methods into this utility class?

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Map;
+import java.util.Set;
+
+import static com.facebook.presto.execution.QueryState.RUNNING;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class TestQueryRunnerUtil
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: Utility classes should be final

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            queryRunner.createCatalog("tpch", "tpch");
+            return queryRunner;
+        }
+        catch (Exception e) {
+            queryRunner.close();
+            throw e;
+        }
+    }
+
+    public static Session newSession()
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: Don't move these methods as they're specific to `TestQueues`

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+import static com.facebook.presto.execution.QueryState.RUNNING;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class TestQueryRunnerUtil
+{
+    private TestQueryRunnerUtil() {}
+
+    public static final String LONG_LASTING_QUERY = "SELECT COUNT(*) FROM lineitem";
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: Don't move this as it's specific to `TestQueues`

File: presto-main/src/test/java/com/facebook/presto/execution/TestRollbackTask.java [hunk 9]
```diff
@@ -68,6 +68,9 @@ public void testRollback()
         Session session = sessionBuilder()
                 .setTransactionId(transactionManager.beginTransaction(false))
                 .build();
+        // Simulate the START TRANSACTION task and set the transaction as inactive
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/execution/TestRollbackTask.java]: Nit: newline before this

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java [hunk 10]
```diff
@@ -108,4 +115,14 @@ public static Session newRejectionSession()
                 .setSource("reject")
                 .build();
     }
+
+    public static Session newTransactionSession()
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: Keep this in the class where it's used

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java [hunk 11]
```diff
@@ -33,12 +34,18 @@
     private TestQueryRunnerUtil() {}
 
     public static final String LONG_LASTING_QUERY = "SELECT COUNT(*) FROM lineitem";
+    public static final String START_TXN_QUERY = "START TRANSACTION";
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: Inline this. Having a constant makes the code harder to read.

File: presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+import static com.facebook.presto.execution.TestQueryRunnerUtil.LONG_LASTING_QUERY;
+import static com.facebook.presto.execution.TestQueryRunnerUtil.START_TXN_QUERY;
+import static com.facebook.presto.spi.ErrorType.USER_ERROR;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_CONCURRENT_QUERY;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+
+public class TestConcurrentQueriesInSingleTransaction
+{
+    @Test(timeOut = 240_000)
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java]: Do we really expect this to run for 4 minutes? This should probably be 1 minute at the most

File: presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    public void testConcurrentQueryInSingleTransaction()
+            throws Exception
+    {
+        try (DistributedQueryRunner queryRunner = TestQueryRunnerUtil.createQueryRunner(ImmutableMap.of())) {
+            queryRunner.installPlugin(new ResourceGroupManagerPlugin());
+
+            Session clientSession = TestQueryRunnerUtil.newTransactionSession();
+
+            // start transaction
+            QueryId startTxnQry = TestQueryRunnerUtil.createQuery(queryRunner, clientSession, START_TXN_QUERY);
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java]: Don't abbreviate "qry"

File: presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+            // submit first "cli" query
+            QueryId firstCliQuery = TestQueryRunnerUtil.createQuery(queryRunner, clientSession, txnId, LONG_LASTING_QUERY);
+
+            // wait for the first query to start
+            TestQueryRunnerUtil.waitForQueryState(queryRunner, firstCliQuery, RUNNING);
+
+            // submit second "cli" query without waiting
+            QueryId secondCliQuery = TestQueryRunnerUtil.createQuery(queryRunner, clientSession, txnId, LONG_LASTING_QUERY);
+
+            // The second query should failed
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java]: nit: start all the comments with lowercase to be consistent

File: presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+            QueryInfo failedQueryInfo = queryRunner.getCoordinator().getQueryManager().getQueryInfo(secondCliQuery);
+            assertNotNull(failedQueryInfo.getErrorType());
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java]: No need for these null checks since the `assertEquals` will do that

File: presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            assertNotNull(failedQueryInfo.getErrorCode());
+            assertEquals(failedQueryInfo.getErrorType(), USER_ERROR);
+            assertEquals(failedQueryInfo.getErrorCode(), INVALID_CONCURRENT_QUERY.toErrorCode());
+
+            // Cancel the first query
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java]: Why is this needed? It should be shutdown automatically when we destroy the query runner

File: presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java (hunk 12 above)
Comment (jessesleeping) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java]: What about other tests (e.g. TestQueues) that have the same 240s time out? I think we can also shorten them to 60s.

File: presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java (hunk 16 above)
Comment (jessesleeping) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java]: If we do it automatically at shutdown, there will be several stack traces in the test log because of some sever exceptions. Adding these two lines simplifies the test log and makes it easier to debug when the test fails.

File: presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+            // submit second "cli" query without waiting
+            QueryId secondCliQuery = createQuery(queryRunner, clientSession, txnId, LONG_LASTING_QUERY);
+
+            // the second query should failed
+            QueryInfo failedQueryInfo = queryRunner.getCoordinator().getQueryManager().getQueryInfo(secondCliQuery);
+            assertEquals(failedQueryInfo.getErrorType(), USER_ERROR);
+            assertEquals(failedQueryInfo.getErrorCode(), INVALID_CONCURRENT_QUERY.toErrorCode());
+
+            // Cancel the first query
+            TestQueryRunnerUtil.cancelQuery(queryRunner, firstCliQuery);
```
Comment (jessesleeping) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestConcurrentQueriesInSingleTransaction.java]: If we do it automatically at shutdown, there will be several stack traces in the test log because of some sever exceptions. Adding these two lines simplifies the test log and makes it easier to debug when the test fails.

**Comment by jessesleeping** [comment]
@electrum @raghavsethi :)

File: presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Map;
+import java.util.Set;
+
+import static com.facebook.presto.execution.QueryState.RUNNING;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public final class TestQueryRunnerUtil
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestQueryRunnerUtil.java]: Name this `QueryRunnerUtil`. We use the `Test` prefix to mean a class that has tests (method with `@Test`) and `Testing` to be an object that is used for tests (similar to a mock).

**Comment by jessesleeping** [comment]
The failed test is related to the flaky test issue described in #7689.
